### ISSUE TITLE
GRPO: sampled prompts repeat across steps, within a global step

### ISSUE THREAD

**Issue by tilaks** [huggingface/trl#3675]
### Reproduction

With gradient_accumulation_steps > 1, `RepeatSampler` generates the same inputs for every step within a global step.
For example, here is a repro using the logic in `RepeatSampler` and `_get_train_sampler`:

```python
from trl.trainer.grpo_trainer import RepeatSampler

# Sample dataset with 9 unique prompts.
dataset = range(9)

# These are the numbers from the example in `_get_train_sampler`
num_generations = 2
generation_batch_size = 6
num_iterations = 1
steps_per_generation = 4

sampler = RepeatSampler(
    data_source=dataset,
    mini_repeat_count=num_generations,
    batch_size=generation_batch_size // num_generations,
    repeat_count=num_iterations * steps_per_generation,
    shuffle=False,
    seed=0,
)
list(sampler)
```

outputs:

```
[0, 0, 1, 1, 2, 2, 0, 0, 1, 1, 2, 2, 0, 0, 1, 1, 2, 2, 0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 3, 3, 4, 4, 5, 5, 3, 3, 4, 4, 5, 5, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7, 8, 8, 6, 6, 7, 7, 8, 8, 6, 6, 7, 7, 8, 8, 6, 6, 7, 7, 8, 8]
```

This may be related to the fact that `RepeatSampler` is not aware of the difference between `num_iterations` and `steps_per_generation`, since it takes a single `repeat_count`.
Therefore, it repeats the chunk verbatim `steps_per_generation` times.
```
for chunk in indexes:
    for _ in range(self.repeat_count):  # i.e. steps_per_generation times in this example
        for index in chunk:
            for _ in range(self.mini_repeat_count):
                yield index
```

### System Info

- Platform: Linux-6.8.0-1026-azure-x86_64-with-glibc2.35
- Python version: 3.10.18
- TRL version: 0.19.0
- PyTorch version: 2.6.0
- accelerator(s): NVIDIA H100 NVL
- Transformers version: 4.53.0
- Accelerate version: 1.4.0
- Accelerate config: not found
- Datasets version: 3.4.1
- HF Hub version: 0.33.1
- bitsandbytes version: 0.46.0
- DeepSpeed version: 0.16.4
- Diffusers version: 0.34.0
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 1.84.0
- PEFT version: 0.15.2
- vLLM version: not installed

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by tilaks** [comment]
@qgallouedec if the current behavior is not as intended, I am happy to put up a PR to modify the behavior.

**Comment by tilaks** [comment]
The default value of `generation_batch_size` includes `* steps_per_generation`.
Therefore, the `batch_size` passed to `RepeatSampler` is reasonable, and afaict. is in line with the batch used by `get_train_dataloader`.
However, I don't understand why `repeat_count` includes `* steps_per_generation`. Doesn't that repeat the same inputs for `steps_per_generation` global steps?

**Comment by tilaks** [comment]
On further reading, `prepare_inputs` has to receive same the `generation_batch` on each step, thus the redundancy.

**Comment by tangyd** [comment]
Hi, have you figured out this problem? I have the same confusion

**Comment by CM-BF** [comment]
> On further reading, `prepare_inputs` has to receive same the `generation_batch` on each step, thus the redundancy.

I still don't think this is a reasonable behavior since `RepeatSampler` really obeys the behavior mentioned in `_get_train_sampler`'s comment string. `steps_per_generation` only works for generate future steps in advance, which has no relationship with repeating. After updating the TRL to the latest version, using this `RepeatSampler` makes my original results not reproducible.

**Comment by yukiumi13** [comment]
> > On further reading, `prepare_inputs` has to receive same the `generation_batch` on each step, thus the redundancy.
> 
> I still don't think this is a reasonable behavior since `RepeatSampler` really obeys the behavior mentioned in `_get_train_sampler`'s comment string. `steps_per_generation` only works for generate future steps in advance, which has no relationship with repeating. After updating the TRL to the latest version, using this `RepeatSampler` makes my original results not reproducible.

I think it is a hack here. For every generation batch (sized per_device_train_batch_size * num_processes * steps_per_generation), it will be used for (steps_per_gen * num_iterations) backward steps in the generate_every cycle, as shown at:
https://github.com/huggingface/trl/blob/e622196097109080b73584d598d4162e64fc6bea/trl/trainer/grpo_trainer.py#L1084
However, to boost efficiency, we want to infer the entire generation batch during the first step of this cycle. Therefore, only the initial (per_device_train_batch_size * num_processes * steps_per_generation) indices should be used, while others are just padding, as the dataloader will be called after each micro backward step. So the total should be (per_device_train_batch_size * num_processes * steps_per_gen) * (step_per_gen * num_iterations).
